src/Microsoft.ML.Transforms/GcnTransform.cs

```diff
@@ -225,477 +255,725 @@ public void Save(ModelSaveContext ctx)
+            base(env, columns)
         {
-            if (count == 0)
-                return 0;
-            return CpuMathUtils.MaxAbsDiff(mean, values.AsSpan(0, count));
+
         }
+    }
 
-        private static Float Mean(Float[] src, int count, int length)
+    /// <summary>
+    /// Extensions for statically typed <see cref="LpNormalizingEstimator"/>.
+    /// </summary>
+    public static class LpNormNormalizerExtensions
+    {
+        private sealed class OutPipelineColumn : Vector<float>
+        {
+            public readonly Vector<float> Input;
+
+            public OutPipelineColumn(Vector<float> input, LpNormEstimatorBase.NormalizerKind normKind, bool subMean)
+                : base(new Reconciler(normKind, subMean), input)
+            {
+                Input = input;
+            }
+        }
+
+        private sealed class Reconciler : EstimatorReconciler
         {
-            if (length == 0 || count == 0)
-                return 0;
-            return CpuMathUtils.Sum(src.AsSpan(0, count)) / length;
+            private readonly LpNormEstimatorBase.NormalizerKind _normKind;
+            private readonly bool _subMean;
+
+            public Reconciler(LpNormEstimatorBase.NormalizerKind normKind, bool subMean)
+            {
+                _normKind = normKind;
+                _subMean = subMean;
+            }
+
+            public override IEstimator<ITransformer> Reconcile(IHostEnvironment env,
+                PipelineColumn[] toOutput,
+                IReadOnlyDictionary<PipelineColumn, string> inputNames,
+                IReadOnlyDictionary<PipelineColumn, string> outputNames,
+                IReadOnlyCollection<string> usedNames)
+            {
+                Contracts.Assert(toOutput.Length == 1);
+
+                var pairs = new List<(string input, string output)>();
+                foreach (var outCol in toOutput)
+                    pairs.Add((inputNames[((OutPipelineColumn)outCol).Input], outputNames[outCol]));
+
+                return new LpNormalizingEstimator(env, pairs.ToArray(), _normKind, _subMean);
+            }
         }
+
+        /// <include file='doc.xml' path='doc/members/member[@name="LpNormalize"]/*'/>
+        /// <param name="input">The column to apply to.</param>
+        /// <param name="normKind">Type of norm to use to normalize each sample.</param>
+        /// <param name="subMean">Subtract mean from each value before normalizing.</param>
+        public static Vector<float> LpNormalize(this Vector<float> input,
+            LpNormEstimatorBase.NormalizerKind normKind = LpNormEstimatorBase.Defaults.NormKind,
+            bool subMean = LpNormEstimatorBase.Defaults.LpSubMean) => new OutPipelineColumn(input, normKind, subMean);
     }
 
-    public static class LpNormalization
+    /// <summary>
+    /// Extensions for statically typed <see cref="GcnNormalizingEstimator"/>.
+    /// </summary>
+    public static class GcNormalizerExtensions
```

Comment:
>GcNormalizerExtensions [](start = 24, length = 22)

all pigstensions should reside in `Microsoft.ML.StaticPipe` or sub-namespaces thereof

Reply:
Randomly clicked on some pigstensions in StaticPipeTests, none of them is in Microsoft.ML.StaticPipe, can you give me example?
I don't mind to move them, I just want to know how process look like.

---
In reply to: [232437263](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232437263) [](ancestors = 232437263)

Reply:
Microsoft.ML.FastTree/TreeTrainersStatic.cs
Microsoft.ML.LightGBM/LightGbmStatic.cs
Microsoft.ML.StandardLearners/Standard/SdcaStatic.cs

---
In reply to: [232770306](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232770306) [](ancestors = 232770306,232437263)

Reply:
see TransformsStatic.cs.

---
In reply to: [232908484](https://github.com/dotnet/machinelearning/pull/1582#discussion_r232908484) [](ancestors = 232908484,232770306,232437263)